src/main/java/com/google/firebase/internal/ApacheHttp2AsyncEntityProducer.java

```diff
@@ -0,0 +1,149 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.internal;
+
+import com.google.api.client.util.StreamingContent;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.nio.AsyncEntityProducer;
+import org.apache.hc.core5.http.nio.DataStreamChannel;
+
+@SuppressWarnings("deprecation")
+public class ApacheHttp2AsyncEntityProducer implements AsyncEntityProducer {
+  private ByteBuffer bytebuf;
+  private ByteArrayOutputStream baos;
+  private final StreamingContent content;
+  private final ContentType contentType;
+  private final long contentLength;
+  private final String contentEncoding;
+  private final CompletableFuture<Void> writeFuture;
+  private final AtomicReference<Exception> exception;
+
+  public ApacheHttp2AsyncEntityProducer(StreamingContent content, ContentType contentType,
+      String contentEncoding, long contentLength, CompletableFuture<Void> writeFuture) {
+    this.content = content;
+    this.contentType = contentType;
+    this.contentEncoding = contentEncoding;
+    this.contentLength = contentLength;
+    this.writeFuture = writeFuture;
+    this.bytebuf = null;
+
+    this.baos = new ByteArrayOutputStream((int) (contentLength < 0 ? 0 : contentLength));
+    this.exception = new AtomicReference<>();
+  }
+
+  public ApacheHttp2AsyncEntityProducer(ApacheHttp2Request request,
+      CompletableFuture<Void> writeFuture) {
+    this(
+        request.getStreamingContent(),
+        ContentType.parse(request.getContentType()),
+        request.getContentEncoding(),
+        request.getContentLength(),
+        writeFuture);
+  }
+
+  @Override
+  public boolean isRepeatable() {
+    return true;
+  }
+
+  @Override
+  public String getContentType() {
+    return contentType != null ? contentType.toString() : null;
+  }
+
+  @Override
+  public long getContentLength() {
+    return contentLength;
+  }
+
+  @Override
+  public int available() {
+    return Integer.MAX_VALUE;
+  }
+
+  @Override
+  public String getContentEncoding() {
+    return contentEncoding;
+  }
+
+  @Override
+  public boolean isChunked() {
+    return contentLength == -1;
+  }
+
+  @Override
+  public Set<String> getTrailerNames() {
+    return null;
+  }
+
+  @Override
+  public void produce(DataStreamChannel channel) throws IOException {
+    if (bytebuf == null) {
+      if (content != null) {
+        try {
+          content.writeTo(baos);
+        } catch (IOException e) {
+          writeFuture.completeExceptionally(e);
+          // failed(e);
```

Comment:
nit: can we remove this now?

Reply:
This should have been uncommented. Fixed that and fixed an issue where exceptions were not being re-thrown.